servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/DefaultCompositeCloseable.java

```diff
@@ -112,39 +125,85 @@ public void closeGracefully() {
     }
 
     private void mergeCloseableDelayError(final AsyncCloseable closeable) {
-        closeAsync = closeAsync.mergeDelayError(closeable.closeAsync());
-        closeAsyncGracefully = closeAsyncGracefully.mergeDelayError(closeable.closeAsyncGracefully());
+        final Operand operand = getOrAddMergeOperand();
+        operand.closables.add(closeable);
+        resetState();
     }
 
     private void mergeCloseableDelayError(final List<AsyncCloseable> closeables) {
-        closeAsync = closeAsync.mergeDelayError(closeables.stream().map(AsyncCloseable::closeAsync).collect(toList()));
-        closeAsyncGracefully = closeAsyncGracefully.mergeDelayError(
-                closeables.stream().map(AsyncCloseable::closeAsyncGracefully).collect(toList()));
+        final Operand operand = getOrAddMergeOperand();
+        operand.closables.addAll(closeables);
+        resetState();
+    }
+
+    private void resetState() {
+        closeAsync = closeAsyncGracefully = null;
+    }
+
+    private Operand getOrAddMergeOperand() {
+        return getOrAddOperand(true, true);
+    }
+
+    private Operand getOrAddConcatOperand(boolean append) {
+        return getOrAddOperand(append, false);
+    }
+
+    private Operand getOrAddOperand(boolean append, boolean isMerge) {
+        final Operand operand;
+        if (operands.isEmpty()) {
+            operand = new Operand(isMerge);
+            operands.add(operand);
+        } else {
+            Operand rawOperand = append ? operands.getLast() : operands.getFirst();
+            if (isMerge == rawOperand.isMerge) {
+                operand = rawOperand;
+            } else {
+                operand = new Operand(isMerge);
+                if (append) {
+                    operands.add(operand);
+                } else {
+                    operands.addFirst(operand);
+                }
+            }
+        }
+        return operand;
     }
 
     private void concatCloseableDelayError(final AsyncCloseable closeable) {
```

Comment:
While you are here, consider renaming `concatCloseableDelayError` -> `appendCloseableDelayError`. We used `concat` before we changed it to append/prepend.